File: WaysOfLooking/W04/Java-Dynamic-Fish/RockPaperScissors.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Scanner;
import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONArray;

public class RockPaperScissors {
    private static JSONObject gameJson;

    static {
        try {
            File jsonFile = new File("C:/Users/auto1/go/src/github.com/eejai42/rock-paper-scissors/SSoT/rps.json");
            Scanner sc = new Scanner(jsonFile);
            String jsonString = "";
            while(sc.hasNextLine()){
                jsonString += sc.nextLine();
            }
            sc.close();
            JSONObject jo = new JSONObject(jsonString);
            gameJson = jo.getJSONObject("rock-paper-scissors");
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        String strategyGuide = GetStrategyGuide();
        int totalScore = CalculateGameScore(strategyGuide);
        System.out.println("Total score: " + totalScore);
    }

     static int CalculateGameScore(String strategyGuide) {
        int totalScore = 0;
        for (int i = 0; i < strategyGuide.length(); i += 2) {
            char opponentChoice = strategyGuide.charAt(i);
            char playerChoice = strategyGuide.charAt(i + 1);
            int roundScore = CalculateRoundScore(opponentChoice, playerChoice);
            System.out.println("Round Score: " + roundScore);
            totalScore += roundScore;
        }
        return totalScore;
    }

     private static int CalculateRoundScore(char opponentChoice, char playerChoice) {
        JSONObject outcomes = gameJson.getJSONObject("rules").getJSONObject("outcomes");
        JSONArray shapes = gameJson.getJSONObject("rules").getJSONArray("shapes");
        JSONObject opponentShape = new JSONObject();
        JSONObject playerShape = new JSONObject();
        for (int i = 0; i < shapes.length(); i++) {
            JSONObject shape = shapes.getJSONObject(i);
            if(shape.getString("opp_code").equals(Character.toString(opponentChoice))) {
                opponentShape = shape;
            }
            if(shape.getString("player_code").equals(Character.toString(playerChoice))){
                playerShape = shape;
            }
        }
        if (playerShape.getString("wins_against").equals(opponentShape.getString("opp_code"))) {
            return outcomes.getInt("win") + playerShape.getInt("score");
        }
        else if (opponentShape.getString("wins_against").equals(playerShape.getString("opp_code"))) {
            return outcomes.getInt("loss") + playerShape.getInt("score");
        }
        else return outcomes.getInt("draw") + playerShape.getInt("score");
    }


    private static String GetStrategyGuide() {
        JSONObject games = gameJson.getJSONObject("rules").getJSONObject("games");
        JSONArray rounds = games.getJSONArray("rounds");
        String strategyGuide = "";
        for (int i = 0; i < rounds.length(); i++) {
            JSONObject round = rounds.getJSONObject(i);
            strategyGuide += round.getString("opp_code") + round.getString("player_code");
        }
        return strategyGuide;
    }
}

File: WaysOfLooking/W09/Fluid-Fish/RockPaperScissors.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
)

type RockPaperScissors struct {
	gameJson map[string]interface{}
}

func NewRockPaperScissors() *RockPaperScissors {
	r := &RockPaperScissors{}

	file, err := ioutil.ReadFile("C:/Users/auto1/go/src/github.com/eejai42/rock-paper-scissors/SSoT/rps.json")
	if err != nil {
		fmt.Println(err)
		return nil
	}

	var data map[string]interface{}
	err = json.Unmarshal(file, &data)
	if err != nil {
		fmt.Println(err)
		return nil
	}

	r.gameJson = data["rock-paper-scissors"].(map[string]interface{})

	return r
}

func  (r *RockPaperScissors)  GetStrategyGuide() string {
    games := r.gameJson["rules"].(map[string]interface{})["games"].(map[string]interface{})
    rounds := games["rounds"].([]interface{})
    var strategyGuide string
    for _, round := range rounds {
        r := round.(map[string]interface{})
        strategyGuide += r["opp_code"].(string) + " " + r["player_code"].(string) + "\n"
    }
    return strategyGuide
}


func (r *RockPaperScissors) CalculateGameScore(strategyGuide string) int {
	totalScore := 0
	for i := 0; i < len(strategyGuide); i += 5 {
		opponentChoice := rune(strategyGuide[i])
		playerChoice := rune(strategyGuide[i+2])
		roundScore := r.CalculateRoundScore(opponentChoice, playerChoice)
		fmt.Println("Round Score: ", roundScore)
		totalScore += roundScore
	}
	return totalScore
}

func (r *RockPaperScissors) CalculateRoundScore(opponentChoice rune, playerChoice rune) int {
	outcomes := r.gameJson["rules"].(map[string]interface{})["outcomes"].(map[string]interface{})
	shapes := r.gameJson["rules"].(map[string]interface{})["shapes"].([]interface{})
	fmt.Printf("SHAPES: %v - %v, %v", shapes, opponentChoice, playerChoice)
	var opponentShape, playerShape map[string]interface{}
	for _, shape := range shapes {
	shapeMap := shape.(map[string]interface{})
	if string(opponentChoice) == shapeMap["opp_code"] {
		opponentShape = shapeMap
	}
	if string(playerChoice) == shapeMap["player_code"] {
		playerShape = shapeMap
	}
}

	fmt.Printf("JSON: %v\nPlayer: %v\nOpp: %v", outcomes, playerShape, opponentShape)

	if opponentShape["wins_against"] == playerShape["opp_code"] {
		return int(outcomes["win"].(float64)) + int(playerShape["score"].(float64))
	} else if playerShape["wins_against"] == opponentShape["opp_code"] {
		return int(outcomes["loss"].(float64)) + int(playerShape["score"].(float64))
	} else {
		return int(outcomes["draw"].(float64)) + int(playerShape["score"].(float64))
	}
}


func (r *RockPaperScissors) Main() int {
	strategyGuide := r.GetStrategyGuide()
	totalScore := r.CalculateGameScore(strategyGuide)
	fmt.Println("Total score: ", totalScore)
	return totalScore
}

func main() {
	r := NewRockPaperScissors()
	r.Main()	
}